File: lib/player-similarity-model.ts
import fs from 'fs'
import path from 'path'
import { parse } from 'csv-parse/sync'
import { StandardScaler } from 'ml-stat'
import KNN from 'ml-knn'

export class PlayerSimilarityModel {
  constructor(dataPath) {
    this.dataPath = dataPath
    this.df = null
    this.transformedData = null
    this.featureNames = null
    this.knnModel = null
  }

  loadAndPreprocessData() {
    const csvData = fs.readFileSync(path.join(process.cwd(), this.dataPath), 'utf-8')
    this.df = parse(csvData, { columns: true, skip_empty_lines: true })
    
    // Feature engineering
    this.df = this.df.map((row, index) => ({
      ...row,
      id: index,
      atk_workrate: row.work_rate.split('/')[0].trim(),
      def_workrate: row.work_rate.split('/')[1].trim(),
    }))

    this._defineFeatureGroups()
  }

  _defineFeatureGroups() {
    this.categoricalFeatures = ['country', 'club', 'best_position']
    this.ordinalFeatures = ['atk_workrate', 'def_workrate']
    this.binaryFeatures = ['preferred_foot']
    this.dropFeatures = ['name', 'first_name', 'last_name', 'work_rate', 'id']

    this.numericFeatures = Object.keys(this.df[0]).filter(
      key => !this.categoricalFeatures.includes(key) &&
             !this.dropFeatures.includes(key) &&
             !this.binaryFeatures.includes(key) &&
             !this.ordinalFeatures.includes(key)
    )
  }

  createPreprocessingPipeline() {
    // This method is not needed in the JavaScript version
    // We'll handle preprocessing in the fitTransformData method
  }

  fitTransformData() {
    const scaler = new StandardScaler()
    
    // Preprocess numeric features
    const numericData = this.df.map(row => this.numericFeatures.map(feature => parseFloat(row[feature])))
    const scaledNumericData = scaler.fit(numericData).transform(numericData)

    // Preprocess categorical features (one-hot encoding)
    const categoricalData = this.categoricalFeatures.flatMap(feature => {
      const uniqueValues = [...new Set(this.df.map(row => row[feature]))]
      return uniqueValues.map(value => this.df.map(row => row[feature] === value ? 1 : 0))
    })

    // Preprocess ordinal features
    const ordinalMap = {
      'High': 2,
      'Medium': 1,
      'Low': 0
    }
    const ordinalData = this.ordinalFeatures.map(feature => 
      this.df.map(row => ordinalMap[row[feature]])
    )

    // Preprocess binary features
    const binaryData = this.binaryFeatures.map(feature =>
      this.df.map(row => row[feature] === 'Right' ? 1 : 0)
    )

    // Combine all preprocessed features
    this.transformedData = scaledNumericData.map((row, i) => [
      ...row,
      ...categoricalData.map(col => col[i]),
      ...ordinalData.map(col => col[i]),
      ...binaryData.map(col => col[i])
    ])

    // Generate feature names
    this.featureNames = [
      ...this.numericFeatures,
      ...this.categoricalFeatures.flatMap(feature => 
        [...new Set(this.df.map(row => row[feature]))].map(value => `${feature}_${value}`)
      ),
      ...this.ordinalFeatures,
      ...this.binaryFeatures
    ]
  }

  fitKnnModel(nNeighbors = 5) {
    this.knnModel = new KNN(this.transformedData, this.df.map(row => row.id), { k: nNeighbors + 1 })
  }

  findSimilarPlayers(playerName) {
    const playerIndex = this.df.findIndex(row => row.name === playerName)
    if (playerIndex === -1) {
      throw new Error(`Player "${playerName}" not found`)
    }

    const playerVector = this.transformedData[playerIndex]
    const neighbors = this.knnModel.predict([playerVector])
    
    const similarPlayers = neighbors[0].slice(1).map(index => {
      const player = this.df[index]
      return {
        name: player.name,
        age: parseInt(player.age),
        overall: parseInt(player.overall),
        potential: parseInt(player.potential),
        value: player.value,
        country: player.country,
        club: player.club,
        best_position: player.best_position,
        preferred_foot: player.preferred_foot,
        similarity_score: 1 - this.cosineSimilarity(playerVector, this.transformedData[index])
      }
    })

    return similarPlayers.sort((a, b) => b.similarity_score - a.similarity_score)
  }

  cosineSimilarity(a, b) {
    const dotProduct = a.reduce((sum, _, i) => sum + a[i] * b[i], 0)
    const magnitudeA = Math.sqrt(a.reduce((sum, val) => sum + val * val, 0))
    const magnitudeB = Math.sqrt(b.reduce((sum, val) => sum + val * val, 0))
    return dotProduct / (magnitudeA * magnitudeB)
  }
}

File: backend/player_similarity_model.py
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.neighbors import NearestNeighbors

class PlayerSimilarityModel:
    def __init__(self, data_path):
        self.data_path = data_path
        self.df = None
        self.transformed_data = None
        self.feature_names = None
        self.column_transformer = None
        self.knn_model = None

    def load_and_preprocess_data(self):
        self.df = pd.read_csv(self.data_path)
        self.df["id"] = self.df.index
        self.df[["atk_workrate", "def_workrate"]] = self.df["work_rate"].str.split("/", expand=True)
        self.df["def_workrate"] = self.df["def_workrate"].str.strip()
        self._define_feature_groups()

    def _define_feature_groups(self):
        self.categorical_features = ["country", "club", "best_position"]
        self.ordinal_features = ["atk_workrate", "def_workrate"]
        self.binary_features = ["preferred_foot"]
        self.drop_features = ["name", "first_name", "last_name", "work_rate", "id"]
        self.numeric_features = list(set(self.df.columns) - 
                                     set(self.categorical_features) - 
                                     set(self.drop_features) -
                                     set(self.binary_features) -
                                     set(self.ordinal_features))

    def create_preprocessing_pipeline(self):
        numeric_transformer = Pipeline([
            ('scaler', StandardScaler())
        ])
        categorical_transformer = Pipeline([
            ('onehot', OneHotEncoder(handle_unknown="ignore"))
        ])
        ordinal_transformer = Pipeline([
            ('ordinal', OneHotEncoder(drop="if_binary"))
        ])
        binary_transformer = Pipeline([
            ('binary', OneHotEncoder(drop="if_binary"))
        ])

        self.column_transformer = ColumnTransformer([
            ('num', numeric_transformer, self.numeric_features),
            ('cat', categorical_transformer, self.categorical_features),
            ('ord', ordinal_transformer, self.ordinal_features),
            ('bin', binary_transformer, self.binary_features)
        ])

    def fit_transform_data(self):
        self.transformed_data = self.column_transformer.fit_transform(self.df)
        self.feature_names = (
            self.numeric_features +
            self.column_transformer.named_transformers_['cat'].named_steps['onehot'].get_feature_names_out(self.categorical_features).tolist() +
            self.column_transformer.named_transformers_['ord'].named_steps['ordinal'].get_feature_names_out(self.ordinal_features).tolist() +
            self.column_transformer.named_transformers_['bin'].named_steps['binary'].get_feature_names_out(self.binary_features).tolist()
        )

    def fit_knn_model(self, n_neighbors=5):
        self.knn_model = NearestNeighbors(
            n_neighbors=n_neighbors + 1,
            metric="cosine",
            algorithm="brute"
        ).fit(self.transformed_data)

    def find_similar_players(self, player_name):
        player_id = self.df[self.df["name"].str.lower() == player_name.lower()]["id"].iloc[0]
        player_vector = self.transformed_data[player_id].reshape(1, -1)
        distances, indices = self.knn_model.kneighbors(player_vector)

        display_cols = ["name", "age", "overall", "potential", "value", 
                        "country", "club", "best_position", "preferred_foot"]
        similar_players = self.df.iloc[indices[0][1:]].reset_index(drop=True)[display_cols]
        similar_players["similarity_score"] = 1 - distances[0][1:]

        return similar_players.sort_values("similarity_score", ascending=False).to_dict('records')

